flink/flink-1.14/src/main/java/io/tidb/bigdata/flink/connector/TiDBDynamicTableFactory.java

```diff
@@ -117,42 +127,217 @@ public DynamicTableSink createDynamicTableSink(Context context) {
 
+        return new JdbcDynamicTableSink(jdbcOptions, jdbcExecutionOptions, jdbcDmlOptions, schema);
+      }
     } else {
       throw new UnsupportedOperationException(
           "Unsupported sink impl: " + tiDBSinkOptions.getSinkImpl());
     }
   }
 
-  private String[] getKeyFields(
+  private void validationForInsertOnDuplicateUpdate(
+      TiDBSinkOptions tiDBSinkOptions, ColumnKeyField columnKeyField, String[] updateColumnNames) {
+    checkArgument(
+        tiDBSinkOptions.getWriteMode() == TiDBWriteMode.UPSERT,
+        "Insert on duplicate only work in `upsert` mode.");
+
+    /**
+     * `UPDATE t1 SET c=c+1 WHERE a=1 OR b=2 LIMIT 1;` If a=1 OR b=2 matches several rows, only one
+     * row is updated. In general, you should try to avoid using an ON DUPLICATE KEY UPDATE clause
+     * on tables with multiple unique indexes.
+     *
+     * @see <a
+     *     href=https://dev.mysql.com/doc/refman/8.0/en/insert-on-duplicate.html>insert-on-duplicate</a>
+     * @see <a href=https://github.com/pingcap/tidb/issues/34813>issue-ON-DUPLICATE-KEY</a>
+     *     <p>The constraints are as follows: - the destination table should contain only one
+     *     not-null unique key(including primary key). - Multiple-Column Indexes should be all
+     *     not-null. - the update columns should contain the unique key column(including primary
+     *     key).
+     */
+    if (!tiDBSinkOptions.isSkipCheckForUpdateColumns()) {
+      List<List<String>> keyFields =
+          Lists.newArrayList(columnKeyField.getUniqueKeys().orElse(Collections.emptyList()));
+      List<String> primaryKey = columnKeyField.getPrimaryKey().orElse(Collections.emptyList());
+      if (!primaryKey.isEmpty()) {
+        keyFields.add(primaryKey);
+      }
+
+      checkArgument(
+          keyFields.size() == 1,
+          "Sink table should only have one unique key or primary key\n"
+              + "If you want to force skip the constraint, "
+              + "set `tidb.sink.skip-check-update-columns` to true");
+
+      TiTableInfo table =
+          columnKeyField
+              .getTableInfo()
+              .orElseThrow(
+                  () ->
+                      new IllegalStateException(
+                          String.format(
+                              "Failed to get tableInfo for table %s.%s",
+                              columnKeyField.getDatabaseName(), columnKeyField.getTableName())));
+      for (String keyField : Objects.requireNonNull(columnKeyField.getKeyFieldFlatMap())) {
+        TiColumnInfo column = table.getColumn(keyField);
+        checkArgument(
+            column.getType().isNotNull(),
+            "Unique key or primary key should be not null\n"
+                + "If you want to force skip the constraint, "
+                + "set `tidb.sink.skip-check-update-columns` to true");
+      }
+
+      ArrayList<String> updateColumns = Lists.newArrayList(updateColumnNames);
+      List<List<String>> collect =
```

Comment:
`keyFields` must have 1 item here. Could we just `checkArgument(updateColumns.containsAll(keyFields[0])`?

Reply:
LGTM